addons/mail/controllers/thread.py

```diff
@@ -150,40 +151,51 @@ def read_subscription_data(self, follower_id):
             ).ids,
         }
 
-    def _prepare_post_data(self, post_data, thread, partner_emails=None, **kwargs):
-        partners = request.env["res.partner"].browse(post_data.pop("partner_ids", []))
+    def _prepare_message_data(self, post_data, *, thread, **kwargs):
+        res = {
+            key: value
+            for key, value in post_data.items()
+            if key in thread._get_allowed_message_params()
+        }
+        if "attachment_ids" in post_data:
+            attachments = request.env["ir.attachment"].browse(
+                [int(aid) for aid in post_data["attachment_ids"]],
+            )
+            if not attachments._has_attachments_ownership(post_data.get("attachment_tokens")):
+                msg = _(
+                    "One or more attachments do not exist, or you do not have the rights to access them.",
+                )
+                raise UserError(msg)
+            res["attachment_ids"] = attachments.ids
         if "body" in post_data:
-            post_data["body"] = Markup(post_data["body"])  # contains HTML such as @mentions
-        if partner_emails:
+            # User input is HTML string, so it needs to be in a Markup.
+            # It will be sanitized by the field itself when writing on it.
+            res["body"] = Markup(post_data["body"]) if post_data["body"] else post_data["body"]
+        partners = request.env["res.partner"].browse(
+            [int(pid) for pid in post_data.get("partner_ids", [])],
+        )
+        if partner_emails := post_data.get("partner_emails"):
             partners |= thread._partner_find_from_emails_single(
                 partner_emails,
                 no_create=not request.env.user.has_group("base.group_partner_manager"),
             )
-        if role_ids := post_data.pop("role_ids", []):
```

Comment:
Note: pop is no longer necessary, as the method no longer blindly forwards non-matching params. `res` is explicitly built and returned instead of changing `post_data` in place and forwarding it.